File: waltid-applications/waltid-wallet-demo-ios/iosApp/iosApp/Components/VerifierReviewSections.swift
import SwiftUI
import WalletSDK

struct VerifierReviewSections: View {
    private enum Request {
        case ready(PresentationRequestInfo)
        case invalid(PresentationRequestContext)
    }

    private let request: Request
    @State private var technicalDetailsExpanded = false

    init(request: PresentationRequestInfo) {
        self.request = .ready(request)
    }

    init(request: PresentationRequestContext) {
        self.request = .invalid(request)
    }

    private var transactionDataGroups: [ClaimGroup] {
        guard case let .ready(request) = request else { return [] }
        return CredentialDisplayNormalizer.transactionDataGroups(for: request)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            if let verifierMetadata,
               verifierDisplayName != nil || !verifierDetails.isEmpty {
                ReviewMetadataSection(
                    title: "Verifier",
                    titleAccessibilityIdentifier: WalletAccessibilityID.presentationVerifierSection
                ) {
                    if let verifierDisplayName {
                        MetadataIdentityView(
                            display: verifierMetadata.display,
                            fallbackName: verifierDisplayName,
                            supportingText: nil
                        )
                        if !verifierDetails.isEmpty {
                            Divider()
                        }
                    }
                    MetadataDetailList(items: verifierDetails)
                }
            }

            ForEach(transactionDataGroups) { group in
                ClaimGroupView(group: group)
            }

            ReviewMetadataSection(
                title: "Response protection",
                titleAccessibilityIdentifier: WalletAccessibilityID.presentationResponseProtectionSection
            ) {
                MetadataDetailList(items: responseProtectionDetails)
            }

            ReviewMetadataSection(
                title: "Technical request details",
                titleAccessibilityIdentifier: WalletAccessibilityID.presentationTechnicalDetailsSection,
                contentInsets: technicalDetailsExpanded
                    ? EdgeInsets(top: 16, leading: 16, bottom: 16, trailing: 16)
                    : EdgeInsets(top: 2, leading: 16, bottom: 2, trailing: 16)
            ) {
                Button {
                    technicalDetailsExpanded.toggle()
                } label: {
                    HStack {
                        Text(technicalDetailsExpanded ? "Hide details" : "Show details")
                        Spacer()
                        Image(systemName: technicalDetailsExpanded ? "chevron.up" : "chevron.down")
                    }
                    .frame(minHeight: 44)
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier(WalletAccessibilityID.verifierTechnicalDetailsToggle)

                if technicalDetailsExpanded {
                    Divider()
                    MetadataDetailList(items: technicalDetails)
                    .accessibilityIdentifier(WalletAccessibilityID.verifierTechnicalDetails)
                }
            }
        }
    }

    private var responseEncryptionStatus: String {
        switch responseEncryption {
        case .notRequired: "Not requested"
        case .required: "Required"
        }
    }

    private var verifierDisplayName: String? {
        guard let name = verifierMetadata?.display?.name?.trimmingCharacters(in: .whitespacesAndNewlines),
              !name.isEmpty else {
            return nil
        }
        return name
    }

    private var verifierDetails: [MetadataDetailItem] {
        guard let metadata = verifierMetadata else { return [] }
        return [
            MetadataDetailItem(label: "Client URI", value: metadata.clientURI, linkURI: metadata.clientURI),
            MetadataDetailItem(label: "Privacy policy", value: metadata.policyURI, linkURI: metadata.policyURI),
            MetadataDetailItem(label: "Terms of service", value: metadata.termsOfServiceURI, linkURI: metadata.termsOfServiceURI),
        ].filter(\.isVisible)
    }

    private var responseProtectionDetails: [MetadataDetailItem] {
        var items = [MetadataDetailItem(label: "Message-level encryption", value: responseEncryptionStatus)]
        if case let .required(details) = responseEncryption {
            items += [
                MetadataDetailItem(label: "Key management algorithm", value: details.keyManagementAlgorithm),
                MetadataDetailItem(label: "Content encryption algorithm", value: details.contentEncryptionAlgorithm),
                MetadataDetailItem(label: "Verifier key ID", value: details.verifierKeyID),
                MetadataDetailItem(label: "Verifier key thumbprint", value: details.verifierKeyThumbprint),
            ]
        }
        return items
    }

    private var technicalDetails: [MetadataDetailItem] {
        [
            MetadataDetailItem(label: "Client ID", value: clientID),
            MetadataDetailItem(label: "Response URI", value: responseURI?.absoluteString),
            MetadataDetailItem(label: "State", value: state),
            MetadataDetailItem(label: "Nonce", value: nonce),
        ]
    }

    private var clientID: String {
        switch request {
        case let .ready(request): request.clientID
        case let .invalid(request): request.clientID
        }
    }

    private var verifierMetadata: VerifierMetadata? {
        switch request {
        case let .ready(request): request.verifierMetadata
        case let .invalid(request): request.verifierMetadata
        }
    }

    private var responseURI: URL? {
        switch request {
        case let .ready(request): request.responseURI
        case let .invalid(request): request.responseURI
        }
    }

    private var state: String? {
        switch request {
        case let .ready(request): request.state
        case let .invalid(request): request.state
        }
    }

    private var nonce: String? {
        switch request {
        case let .ready(request): request.nonce
        case let .invalid(request): request.nonce
        }
    }

    private var responseEncryption: PresentationResponseEncryption {
        switch request {
        case let .ready(request): request.responseEncryption
        case let .invalid(request): request.responseEncryption
        }
    }
}

File: waltid-applications/waltid-wallet-demo-compose/sharedUI/src/commonMain/kotlin/id/walt/walletdemo/compose/ui/components/VerifierReviewSections.kt
package id.walt.walletdemo.compose.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import id.walt.walletdemo.compose.logic.WalletDemoPresentationRequestInfo
import id.walt.walletdemo.compose.logic.WalletDemoResponseEncryption
import id.walt.walletdemo.compose.ui.WalletUiTestTags

@Composable
internal fun VerifierReviewSections(request: WalletDemoPresentationRequestInfo, modifier: Modifier = Modifier) {
    var technicalDetailsExpanded by rememberSaveable { mutableStateOf(false) }
    val verifierMetadata = request.verifierMetadata

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        verifierMetadata?.let { metadata ->
            val displayName = metadata.display?.name?.trim()?.takeIf { it.isNotEmpty() }
            val details = listOf(
                MetadataDetailItem("Client URI", metadata.clientUri, metadata.clientUri),
                MetadataDetailItem("Privacy policy", metadata.policyUri, metadata.policyUri),
                MetadataDetailItem("Terms of service", metadata.termsOfServiceUri, metadata.termsOfServiceUri),
            ).filter { !it.value.isNullOrBlank() }

            if (displayName != null || details.isNotEmpty()) {
                ReviewMetadataSection(
                    title = "Verifier",
                    modifier = Modifier.testTag(WalletUiTestTags.PresentationVerifierSection),
                ) {
                    if (displayName != null) {
                        MetadataIdentityRow(
                            display = metadata.display,
                            fallbackName = displayName,
                            supportingText = null,
                        )
                        if (details.isNotEmpty()) MetadataRowDivider()
                    }
                    MetadataDetailList(details)
                }
            }
        }

        request.transactionData.forEach { group ->
            ClaimGroupSection(group)
        }

        ReviewMetadataSection(
            title = "Response protection",
            modifier = Modifier.testTag(WalletUiTestTags.PresentationResponseProtectionSection),
        ) {
            val encryption = request.responseEncryption
            MetadataDetailList(
                buildList {
                    add(
                        MetadataDetailItem(
                            "Message-level encryption",
                            when (encryption) {
                                WalletDemoResponseEncryption.NotRequired -> "Not requested"
                                is WalletDemoResponseEncryption.Required -> "Required"
                            },
                        )
                    )
                    if (encryption is WalletDemoResponseEncryption.Required) {
                        add(MetadataDetailItem("Key management algorithm", encryption.keyManagementAlgorithm))
                        add(MetadataDetailItem("Content encryption algorithm", encryption.contentEncryptionAlgorithm))
                        add(MetadataDetailItem("Verifier key ID", encryption.verifierKeyId))
                        add(MetadataDetailItem("Verifier key thumbprint", encryption.verifierKeyThumbprint))
                    }
                }
            )
        }

        ReviewMetadataSection(
            title = "Technical request details",
            modifier = Modifier.testTag(WalletUiTestTags.PresentationTechnicalDetailsSection),
            contentPadding = if (technicalDetailsExpanded) {
                PaddingValues(16.dp)
            } else {
                PaddingValues(horizontal = 16.dp, vertical = 2.dp)
            },
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 44.dp)
                    .clickable { technicalDetailsExpanded = !technicalDetailsExpanded }
                    .testTag(WalletUiTestTags.VerifierTechnicalDetailsToggle),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(if (technicalDetailsExpanded) "Hide details" else "Show details")
                Icon(
                    imageVector = if (technicalDetailsExpanded) Icons.Default.KeyboardArrowUp else Icons.Default.KeyboardArrowDown,
                    contentDescription = null,
                )
            }

            if (technicalDetailsExpanded) {
                MetadataRowDivider()
                MetadataDetailList(
                    listOf(
                        MetadataDetailItem("Client ID", request.clientId),
                        MetadataDetailItem("Response URI", request.responseUri),
                        MetadataDetailItem("State", request.state),
                        MetadataDetailItem("Nonce", request.nonce),
                    ),
                    modifier = Modifier.testTag(WalletUiTestTags.VerifierTechnicalDetails),
                )
            }
        }
    }
}
